modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java

```diff
@@ -420,130 +421,51 @@ public void validate(
-				results = objectValidationRuleEngine.execute(
-					(Map<String, Object>)variables.get("entryDTO"), null);
-			}
-
-			Locale locale = LocaleUtil.getMostRelevantLocale();
-
-			User user = _userLocalService.fetchUser(userId);
-
-			if (user != null) {
-				locale = user.getLocale();
-			}
-
-			String errorMessage = null;
-
-			if (!GetterUtil.getBoolean(results.get("validationCriteriaMet"))) {
-				errorMessage = objectValidationRule.getErrorLabel(locale);
-			}
-			else if (GetterUtil.getBoolean(results.get("invalidScript"))) {
-				errorMessage = _language.get(
-					locale, "there-was-an-error-validating-your-data");
-			}
-
-			if (Validator.isNull(errorMessage)) {
-				continue;
-			}
-
-			if (objectValidationRule.compareOutputType(
-					ObjectValidationRuleConstants.
-						OUTPUT_TYPE_PARTIAL_VALIDATION)) {
-
-				for (ObjectValidationRuleSetting objectValidationRuleSetting :
-						_objectValidationRuleSettingPersistence.findByOVRI_N(
-							objectValidationRule.getObjectValidationRuleId(),
-							ObjectValidationRuleSettingConstants.
-								NAME_OUTPUT_OBJECT_FIELD_ID)) {
-
-					ObjectField objectField =
-						_objectFieldPersistence.fetchByPrimaryKey(
-							GetterUtil.getLong(
-								objectValidationRuleSetting.getValue()));
+		_validate(
+			baseModel, objectDefinitionId, payloadJSONObject, userId,
+			objectValidationRules);
+	}
 
-					if (objectField == null) {
-						continue;
-					}
+	@Override
+	@Transactional(readOnly = true)
+	public void validate(
+			ObjectEntry objectEntry, List<String> validationKeys, long userId)
+		throws PortalException {
 
-					objectValidationRuleResults.add(
-						new ObjectValidationRuleResult(
-							errorMessage, objectField.getName()));
-				}
-			}
-			else {
-				objectValidationRuleResults.add(
-					new ObjectValidationRuleResult(errorMessage));
-			}
+		if (validationKeys.isEmpty()) {
+			List<ObjectValidationRule> allObjectValidationRules =
+				objectValidationRuleLocalService.getObjectValidationRules(
+					objectEntry.getObjectDefinitionId());
+
+			_validate(
+				objectEntry, objectEntry.getObjectDefinitionId(),
+				ObjectEntryUtil.getPayloadJSONObject(
+					_dtoConverterRegistry, _jsonFactory, null,
+					_objectDefinitionPersistence.fetchByPrimaryKey(
+						objectEntry.getObjectDefinitionId()),
+					objectEntry, objectEntry, null,
+					_userLocalService.getUser(userId)),
+				userId, allObjectValidationRules);
 		}
-
-		ObjectEntryThreadLocal.addValidatedObjectEntryId(
-			(long)baseModel.getPrimaryKeyObj());
-
-		if (ListUtil.isNotEmpty(objectValidationRuleResults)) {
-			throw new ObjectValidationRuleEngineException(
-				objectValidationRuleResults);
+		else {
+			List<ObjectValidationRule> objectValidationRules =
+				TransformUtil.transform(
+					validationKeys,
+					validationKey ->
+						objectValidationRuleLocalService.
+							fetchObjectValidationRule(
+								validationKey,
+								objectEntry.getObjectDefinitionId()));
+
+			_validate(
```

Comment:
It would be better to call `_validate` method only once, so the idea is to declare `List<ObjectValidationRule> objectValidationRules = null;` first and then populate this variable in different ways in if else and finally call `_validate` after if else condition. Sounds good?